### ISSUE TITLE
current_console_user_id incorrect after another user logs in in background

### ISSUE THREAD

**Issue by Sesquipedalian** [tekezo/Karabiner-Elements#1770]
I've noticed that if another user logs in via a background process such as Screen Sharing, karabiner-grabber can get confused about which UID to use for current_console_user_id.

I don't know exactly how Karabiner keeps track of which user currently controls the GUI, but the most reliable way is simply to check which user currently owns /dev/console. If you were to check the ownership of the console device periodically and take action if the owner had changed, this issue would be easy to solve.

I suspect that a change like this would also fix the root cause of #482 and #1137.

**Comment by tekezo** [comment]
Thank you for feedback!

Surely, Karabiner-Elements is using `SCDynamicStoreCopyConsoleUser` to determine the current console user.
I'll fix the method with suitable way.

https://github.com/tekezo/Karabiner-Elements/blob/bd8c7af4f078746581d30a534fee1c8db581871a/src/vendor/cget/cget/pkg/pqrs-org__cpp-osx-session/install/include/pqrs/osx/session/session.hpp#L15

**Comment by Sesquipedalian** [comment]
Thanks @tekezo. According to Apple's documentation, `SCDynamicStoreCopyConsoleUser` does seem like it should be a reliable way to get this information, so my guess is that Karabiner is sometimes not detecting changes to this value.

**Comment by Sesquipedalian** [comment]
... But I don't really know. 🙂

**Comment by tekezo** [comment]
# Progress
I confirmed `SCDynamicStoreCopyConsoleUser` is wrong, but also `/dev/console` is not enough.
We need more reliable method.

# Steps to reproduce the issue

1. Log in from console as Guest user. (uid 201)

```text
crw-------  1 Guest  _guest    0,   0 Mar 19 09:13 /dev/console
SCDynamicStoreCopyConsoleUser: 201

(Both results are correct.)
```

2. Then log in from Screen Sharing as another account (uid 501)

```text
crw-------  1 tekezo  staff    0,   0 Mar 19 09:13 /dev/console
SCDynamicStoreCopyConsoleUser: 501

(Both results are wrong.)
```

3. Then log out uid 501.

```text
crw-------  1 root  wheel    0,   0 Mar 19 09:16 /dev/console
SCDynamicStoreCopyConsoleUser: 501

(/dev/console is correct, but SCDynamicStoreCopyConsoleUser is wrong.)
```

4. Then log out Guest.

```text
crw-------  1 root  wheel    0,   0 Mar 19 09:17 /dev/console
SCDynamicStoreCopyConsoleUser: empty

(Both results are correct.)
```

**Comment by tekezo** [comment]
This issue will be fixed at v12.4.0 (the next next release).

**Comment by Sesquipedalian** [comment]
Lovely. Thank you. 😁

**Comment by tekezo** [comment]
I'll release v12.4.0 soon as a hotfix for an issue of v12.3.0.

The state of this issue requires more time to fix.
So, I'll fix this issue at v12.5.0.

**Comment by tekezo** [comment]
I've fixed this issue at v12.4.6. (latest beta version)

How to upgrade to the latest beta version:
<https://pqrs.org/osx/karabiner/document.html#check-for-updates>

**Comment by Sesquipedalian** [comment]
Thanks!
